src/wutil/gettext.rs

```diff
@@ -54,114 +256,226 @@ mod gettext_impl {
+fn get_language_var(vars: &EnvStack) -> Option<Vec<String>> {
+    let langs = vars.get_unless_empty(L!("LANGUAGE"))?;
+    let langs = langs.as_list();
+    let filtered_langs: Vec<String> = langs
+        .iter()
+        .filter(|lang| !lang.is_empty())
+        .map(|lang| lang.to_string())
+        .collect();
+    if filtered_langs.is_empty() {
+        return None;
     }
+    Some(filtered_langs)
 }
 
 /// Call this when one of `LANGUAGE`, `LC_ALL`, `LC_MESSAGES`, `LANG` changes.
 /// Updates internal state such that the correct localizations will be used in subsequent
 /// localization requests.
+///
+/// For deciding how to localize, the following is done:
+///
+/// 1. Check the first non-empty value of the env vars `LC_ALL`, `LC_MESSAGES`, `LANG`. If it
+///    starts with `C` we consider this a C locale and disable localization.
+/// 2. Otherwise, if the language precedence was set via `status locale`, env vars are ignored.
+/// 3. Otherwise, the value of the `LANGUAGE` env var is used, if non-empty. This allows specifying
+///    multiple languages, with languages specified first taking precedence, e.g.
+///    `LANGUAGE=zh_TW:zh_CN:pt_BR`
+/// 4. Otherwise, the first non-empty value of the env vars `LC_ALL`, `LC_MESSAGES`, `LANG` is
+///    used. This can only specify a single language, e.g. `LANG=de_AT.UTF-8`.
+///    There, we normalize locale names by stripping off the suffix, leaving only the `ll_CC` part.
+/// 5. Otherwise, localization will not happen.
+///
+/// If users specify `ll_CC` as a language and we don't have a catalog for this language, but we
+/// have one for `ll`, that will be used instead. If users specify `ll` (without specifying a
+/// language variant), which we discourage, and we don't have a catalog for `ll`, but we do have
+/// one for `ll_CC`, that will be used as a fallback. If we have multiple `ll_*` catalogs, all of
+/// them will be used, in arbitrary order.
+#[cfg(feature = "localize-messages")]
+pub fn update_from_env(vars: &EnvStack) {
+    gettext_impl::update_from_env(get_message_locale(vars), get_language_var(vars));
+}
+
+#[cfg(feature = "localize-messages")]
+pub struct SetLocaleLints<'a> {
+    malformed: Vec<&'a str>,
+    duplicates: Vec<&'a str>,
+    non_existing: Vec<&'a str>,
+}
+
+#[cfg(feature = "localize-messages")]
+impl<'a> From<gettext_impl::SetLocaleLints<'a>> for SetLocaleLints<'a> {
+    fn from(lints: gettext_impl::SetLocaleLints<'a>) -> Self {
+        Self {
+            malformed: lints.malformed,
+            duplicates: lints.duplicates,
+            non_existing: lints.non_existing,
+        }
+    }
+}
+
+#[cfg(feature = "localize-messages")]
+impl<'a> SetLocaleLints<'a> {
+    pub fn display_malformed(&self) -> WString {
+        if self.malformed.is_empty() {
+            return WString::new();
+        }
+        wgettext_fmt!(
+            "Language specifiers are not well-formed: %s\n",
+            format!("{:?}", self.malformed)
```

Comment:
this looks like the sort of place where good plural handling in localization will come in handy